### PR TITLE
Re-enable automated CSRF Protection: Add custom XSRF interceptor which works with cross origin requests

### PR DESCRIPTION
**Pull request by tdonohue**
## References
* Should be tested with https://github.com/DSpace/DSpace/pull/2922 (Backend PR which re-enables CSRF protection in Server Webapp).  This PR will do nothing until the backend behavior is modified in 2922.

## Description
This PR creates a customized `XsrfInterceptor` which accepts XSRF/CSRF Cookies from different origins & responds with an X-XSRF-TOKEN header.  The default, Angular `HttpXsrfInterceptor` ONLY works if Angular and your backend are installed on the same origin/server.

## Instructions for Reviewers
* Install the backend PR locally: https://github.com/DSpace/DSpace/pull/2922   This PR must be tested with that Backend PR, as it otherwise does "nothing" (the new Interceptor will never find an `XSRF-TOKEN` Cookie and therefore never respond with an `X-XSRF-TOKEN` Header.
* Test this PR alongside that one to ensure the Angular UI still has the same behavior.  Especially test authentication and authorization (e.g. as an Admin, try to perform an Admin task).
* (Optionally) you may use Chrome Dev Tools (or similar) to verify that an X-XSRF-TOKEN header is sent on non-GET requests (e.g. login sends a POST to the /login endpoint & will include this header)

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs for any bug fixes, improvements or new features. A few reminders about what constitutes good tests:
    * Include tests for different user types (if behavior differs), including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for error scenarios, e.g. when errors/warnings should appear (or buttons should be disabled).
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Comment by tdonohue** [comment]
@artlowel : I've assigned this to you for review. However, feel free to delegate it to someone else on your team as needed.  The code should be easy to review...it's just testing that takes time (as it requires installing the backend somewhere).

File: src/app/core/xsrf/xsrf.interceptor.spec.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    expect(httpRequest.request.headers.has('X-XSRF-TOKEN'));
+    const token = httpRequest.request.headers.get('X-XSRF-TOKEN');
+    expect(token).toBeDefined();
+    expect(token).toBe(testToken.toString());
+  });
+
+  it('should NOT add an X-XSRF-TOKEN header when we’re sending an HTTP GET request', () => {
+    service.request(RestRequestMethod.GET, 'server/api/core/items').subscribe((response) => {
+      expect(response).toBeTruthy();
```
Comment (artlowel) [inline src/app/core/xsrf/xsrf.interceptor.spec.ts:62]: An expect inside a subscribe may not be taken in to account as it is executed asynchronously and that may not happen before the `it` is completed. To be sure, please include a 'done' callback as explained in the [testing guide on the wiki](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide#CodeTestingGuide-WritingUnit(spec)Tests)

File: src/app/core/xsrf/xsrf.interceptor.spec.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    const httpRequest = httpMock.expectOne(`server/api/core/items`);
+
+    expect(!httpRequest.request.headers.has('X-XSRF-TOKEN'));
+  });
+
+  it('should NOT add an X-XSRF-TOKEN header when we’re sending an HTTP POST to an untrusted URL', () => {
+    // POST to a URL which is NOT our REST API
+    service.request(RestRequestMethod.POST, 'https://untrusted.com', 'test').subscribe((response) => {
+      expect(response).toBeTruthy();
```
Comment (artlowel) [inline src/app/core/xsrf/xsrf.interceptor.spec.ts:73]: Same comment

**Review by artlowel — APPROVED** [review]
Thanks @tdonohue!
As discussed, I did a code review only. 
It looks good apart from a few inline comments about the tests

File: src/app/core/xsrf/xsrf.interceptor.spec.ts (hunk 1 above)
Comment (tdonohue) [reply src/app/core/xsrf/xsrf.interceptor.spec.ts:62]: Fixed in https://github.com/DSpace/dspace-angular/pull/833/commits/9bb982043a5f95dcc16993d9bf72aee282d8b97f

File: src/app/core/xsrf/xsrf.interceptor.spec.ts (hunk 2 above)
Comment (tdonohue) [reply src/app/core/xsrf/xsrf.interceptor.spec.ts:74]: Fixed in https://github.com/DSpace/dspace-angular/pull/833/commits/9bb982043a5f95dcc16993d9bf72aee282d8b97f

**Comment by tdonohue** [comment]
@artlowel : After adding in the missing `done()` statements into my tests, they all oddly timeout (and I'm not quite sure why).  

I had initially modeled my `xsrf.interceptor.spec.ts` tests on those in these classes (and I'm noticing none of them have `done()` either):
* https://github.com/DSpace/dspace-angular/blob/main/src/app/core/auth/auth.interceptor.spec.ts
* https://github.com/DSpace/dspace-angular/blob/main/src/app/core/locale/locale.interceptor.spec.ts

I'm at a bit of a loss on how to proceed here. Any ideas on what could be going on?  (No major rush as this PR still needs testing from others, but if anything comes to mind, I'd appreciate more ideas.)

File: src/app/core/xsrf/xsrf.interceptor.spec.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  it('should NOT add an X-XSRF-TOKEN header when we are sending an HTTP POST to an untrusted URL', (done) => {
+    // POST to a URL which is NOT our REST API
+    service.request(RestRequestMethod.POST, 'https://untrusted.com', 'test').subscribe((response) => {
+      expect(response).toBeTruthy();
+      done();
+    });
+
+    const httpRequest = httpMock.expectOne(`https://untrusted.com`);
+
+    expect(!httpRequest.request.headers.has('X-XSRF-TOKEN'));
```
Comment (artlowel) [inline src/app/core/xsrf/xsrf.interceptor.spec.ts:81]: Same as above

File: src/app/core/xsrf/xsrf.interceptor.spec.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+  it('should NOT add an X-XSRF-TOKEN header when we are sending an HTTP GET request', (done) => {
+    service.request(RestRequestMethod.GET, 'server/api/core/items').subscribe((response) => {
+      expect(response).toBeTruthy();
+      done();
+    });
+
+    const httpRequest = httpMock.expectOne(`server/api/core/items`);
+
+    expect(!httpRequest.request.headers.has('X-XSRF-TOKEN'));
```
Comment (artlowel) [inline src/app/core/xsrf/xsrf.interceptor.spec.ts:69]: This expect is incomplete, it needs to be followed by toBe _Something_  in order to do anything. E.g.
```suggestion
    expect(httpRequest.request.headers.has('X-XSRF-TOKEN')).toBeFalse();
```

**Review by artlowel — COMMENTED** [review]
@tdonohue I had to experiment a bit, because I didn't have any experience with HttpTestingController either.

It turns out you have to call `flush()` on the request after the `expectOne` call of each test if you need it to return a response. The param for flush should be the data of the response so that doesn't matter too much in this case.

Note that this is not a universal thing for all async angular tests, but rather something specific to the way this HttpTestingController was designed.

If other existing tests are using the same method, without including done (or another way of dealing with an async expect), it is likely that their expects will be ignored as well.

**Comment by tdonohue** [comment]
Thanks @artlowel !  With your advice, I think I've fixed the issues (at least all tests now succeed locally, we'll see what Travis thinks).   See https://github.com/DSpace/dspace-angular/pull/833/commits/2e99ae4caa5d8395e4f5e4dcc5eee3c5f5364e8d

**Review by atarix83 — COMMENTED** [review]
thanks @tdonohue for this PR
I've deployd rest [PR](https://github.com/DSpace/DSpace/pull/2922) and this one on a test server. An even if it seems to work fine (with both password and shibboleth authentication), when I try from my localhost to login I can't (for both password and shibboleth authentication).
After a quick debug the XsrfInterceptor can't get the xsrf token from cookie maybe I suppose that it's blocked because I'm not on the same rest server's host. I didn't tried but I think this could happens also when angular application and rest server are on different host (the situation we have now on the demo)

**Comment by tdonohue** [comment]
@atarix83 : This CSRF protection should work across hosts as long as you have CORS also configured properly (my suspicion is you should have seen a CORS error somewhere when the login failed -- but if you saw a different error, please let me know).  Assuming it was a CORS error, then make sure the backend's `rest.cors.allowed-origins` includes your frontend UI AND that the `rest.cors.allow-credentials = true` (which is the default setting).  It's the `allow-credentials` setting which enables Cookies in CORS requests.

If I'm misunderstanding the error you saw, let me know. I highly suspect it was a CORS related issue though, but if you saw some different error (either from the frontend or backend), let me know.

**Comment by atarix83** [comment]
@tdonohue 
unfortunately all configurations are right.
However I found a way to reproduce the problem also running angular and rest on my local machine.
The trick is to run the local angular application with an hostnames different form `localhost`.
I've ubuntu as OS so I've added the following line to my `/etc/hosts`  (something similar can be done on windows too) 
```
127.0.0.1       dspace.angular.org
```
the you need to run angular application with the following command (otherwise you get an `Invalid Host header` error) : 
```
ng serve --host 0.0.0.0 -port 4000 --disable-host-check
```
Lastly if you try to login with password you should reproduce the problem

**Comment by tdonohue** [comment]
@atarix83 : Thanks to your example above, I'm able to reproduce the issue by creating a "fake" separate host for the Angular app.  The problem (according to Chrome DevTools) is with the `XSRF-TOKEN` cookie.  Specifically, when looking at the `XSRF-TOKEN` cookie in the `/login` 403 response (on the "Cookies" tab), I see this message: 
> This Set-Cookie didn't specify a "SameSite" attribute and was defaulted to "SameSite=Lax," and was blocked because it came from a cross-site response which was not the response to a top-level navigation. The Set-Cookie had to have been set with "SameSite=None" to enable cross-site usage.


So, it sounds to me like this is the SameSite Cookie issue that you and @abollini reported about a week ago.  I think **ALL** of our authentication/CSRF Cookies in DSpace need to have `SameSite=None; Secure` set so that they can work in cross-site requests.

Since I can now reproduce this locally, I'll look at how to fix this for the XSRF-TOKEN cookie. Once I have a fix, I can also apply the same settings to the DSpace Authorization/Authentication Cookie, which might fix the SameSite Cookie issues you were seeing with Shibboleth on the demo site.

**Comment by tdonohue** [comment]
Closing as I've found major flaws with using Spring Security's default CSRF protection for DSpace 7.  I think it would unfortunately require that all clients of the REST API (including the Angular UI) run on the same primary domain as the REST API.  This seems like it'd not align with our REST API use cases, as it would limit third party integrations.

Much more info (including a new approach) in this comment on the corresponding backend PR: https://github.com/DSpace/DSpace/pull/2922#issuecomment-701531443

**Pull request closed without merging**